bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCache.java

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.cache;
+
+import java.util.function.Function;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a simple cache implementation of a single key-value-pair.
+ *
+ * There must be provided an action in order to retrieve/calculate the value for any given key. This action will be
+ * called only if the answer from the last calculation is not valid anymore, i.e. if the key is different or the last
+ * result is expired.
+ *
+ * @author Simon Kaufmann - Initial contribution and API.
+ * @author Christoph Weitkamp - Moved ExpiringCache from Yahoo Weather Binding to Eclipse SmartHome Core.
+ *
+ * @param <Q> the type of the key
+ * @param <V> the type of the value
+ */
+public class ExpiringCache<Q, V> {
+
+    private final Logger logger = LoggerFactory.getLogger(ExpiringCache.class);
+
+    private final long expiry;
+    private final Function<Q, V> action;
+    private long lastUpdate;
+    private Q lastQuery;
+    private V lastValue;
+
+    /**
+     * Create a new instance.
+     *
+     * @param expiry duration in milliseconds for how long the value stays valid
+     * @param action the action to retrieve/calculate the value
+     */
+    public ExpiringCache(long expiry, Function<Q, V> action) {
+        if (action == null) {
+            throw new IllegalArgumentException(
+                    "ExpiringCache cannot be created as action is not an instance of Function.");
```

Comment:
Hm, why not log what the check is exactly about e.g. `ExpiringCache cannot be created using a null action`.

Sure, we could state that if action is a function (correct argument type), then the only possibility is a null one...